File: markdown-processor/src/test/java/org/bsc/markdown/Issue189Test.java
package org.bsc.markdown;

import static org.hamcrest.core.IsEqual.equalTo;
import static org.hamcrest.core.IsNull.nullValue;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.fail;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.bsc.markdown.PegdownParse;
import org.junit.Test;


/**
 * 
 * @author bsorrentino
 *
 */
public class Issue189Test extends PegdownParse {

    @Override
    protected char[] loadResource() throws IOException {
        return new StringBuilder()
            .append( "![alt text](./images/img.png \"title\")" ).append("\n\n")
            .append( "![](./images/img.png \"title\")" ).append("\n\n")
            .append( "![](./images/img.png \"\")" ).append("\n\n")
            .append( "![](./images/img.png)" ).append("\n\n")
            .append( "![alt text](${pageTitle}^img.png \"title\")" ).append("\n\n")
            .append( "![alt text](img.png \"title\")" ).append("\n\n")
            .append( "![thumbnail](meal.png \"Meal\")" ).append("\n\n")
            .append( "![conf-icon](http://www.lewe.com/wp-content/uploads/2016/03/conf-icon-64.png \"My conf-icon\")" ).append("\n\n")
            .append( "![cb-integration-components.png](cb-integration-components.png)").append("\n\n")
            .toString()
            .toCharArray()
        ;
        
    }
    
    @Test
    public void parse() throws IOException {
        final String [] lines = serializeToString().split( "\n" );
        
        int i=0;
        assertThat( lines[i++], equalTo("!${page.title}^img.png|alt text!"));
        assertThat( lines[i++], equalTo("!${page.title}^img.png|!"));
        assertThat( lines[i++], equalTo("!${page.title}^img.png|!"));
        assertThat( lines[i++], equalTo("!${page.title}^img.png|!"));
        assertThat( lines[i++], equalTo("!${pageTitle}^img.png|alt text!"));
        assertThat( lines[i++], equalTo("!${page.title}^img.png|alt text!"));
        assertThat( lines[i++], equalTo("!${page.title}^meal.png|thumbnail!"));
        assertThat( lines[i++], equalTo("!http://www.lewe.com/wp-content/uploads/2016/03/conf-icon-64.png|conf-icon!"));
        assertThat( lines[i++], equalTo("!${page.title}^cb-integration-components.png|cb-integration-components.png!"));
    }

    /**
     * 
     * @param uri
     * @return
     */
    private CompletableFuture<String> getFileName( String uri ) {
        
        final CompletableFuture<String> result = new CompletableFuture<>();
        try {
            final java.net.URI uriObject = java.net.URI.create(uri);
            
            if( "classpath".equalsIgnoreCase(uriObject.getScheme()) ) {
                result.completeExceptionally( 
                        new IllegalArgumentException( "'classpath' scheme is not supported!"));    
                return result;
            }
            final Path path = Paths.get(uriObject.getPath());
            
            result.complete(path.getFileName().toString());
        
        } catch( Throwable e) {
            
            result.completeExceptionally(e);
        }
        return result;
    }
    
    @Test
    public void processUrl() {
        final Pattern patternUrl = Pattern.compile("(?:(\\$\\{.+\\})\\^)?(.+)");
        {
        final String url = "${page.title}^image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), equalTo("${page.title}") );
        assertThat( m.group(2), equalTo("image-name.png") );
        }
        
        {
        final String url = "${page.title}^./images/image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), equalTo("${page.title}") );
        assertThat( m.group(2), equalTo("./images/image-name.png") );
        
        getFileName(m.group(2))
        .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
        .thenAccept( fileName -> assertThat( fileName, equalTo("image-name.png")))
        .join()
        ;

        }
        
        {
        final String url = "image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
            .thenAccept( fileName -> assertThat( fileName, equalTo("image-name.png")))
            .join()
            ;
        }
        
        {
        final String url = "./images/image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("./images/image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
            .thenAccept( fileName -> assertThat( fileName, equalTo("image-name.png")))
            .join()
            ;
        }
        
        {
        final String url = "file:///image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("file:///image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
            .thenAccept( fileName -> assertThat( fileName, equalTo("image-name.png")))
            .join()
            ;
        }
        
        {
        final String url = "http://localhost:8080/image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("http://localhost:8080/image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
            .thenAccept( fileName -> assertThat( fileName, equalTo("image-name.png")))
            .join()
            ;
        }
        
        {
        final String url = "classpath:/image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("classpath:/image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { return null ;} )
            .thenAccept( fileName -> assertThat( fileName, nullValue()) )
            .join()
            ;
        }
        
        {
        final String url = "/image-name.png";
        final Matcher m = patternUrl.matcher(url);
        
        assertThat( m.matches(), equalTo(true) );
        assertThat( m.groupCount(), equalTo(2) );
        assertThat( m.group(1), nullValue() );
        assertThat( m.group(2), equalTo("/image-name.png") );
        
        getFileName(m.group(2))
            .exceptionally( ex -> { fail( ex.getMessage()); return null ;} )
            .thenApply( fileName -> "${page.title}^".concat(fileName) )
            .thenAccept( fileName -> assertThat( fileName, equalTo("${page.title}^image-name.png")))
            .join()
            ;
        }
        
    }
    
}

File: markdown-processor-commonmark/src/test/kotlin/org/bsc/makdown/commonmark/Issue189Test.kt
package org.bsc.makdown.commonmark

import org.bsc.confluence.model.Site
import org.hamcrest.core.IsEqual
import org.hamcrest.core.IsNull
import org.junit.Assert
import org.junit.Test
import org.junit.jupiter.api.Assertions
import java.io.IOException
import java.net.URI
import java.nio.file.Paths
import java.util.concurrent.CompletableFuture
import java.util.regex.Pattern

/**
 *
 * @author bsorrentino
 */
class Issue189Test  {

    var site = Site().apply {
        basedir = Paths.get(System.getProperty("user.dir"))
    }

    @Test
    fun parse() {
        val content = parseResource( this.javaClass, "issue189", this.site )
        val pageTitle = "\${page.title}"

        Assertions.assertEquals("""
        !${pageTitle}^img.png|alt text!
        
        !${pageTitle}^img.png|!
        
        !${pageTitle}^img.png|!
        
        !${pageTitle}^img.png|!
        
        !${'$'}{pageTitle}^img.png|alt text!
        
        !${pageTitle}^img.png|alt text!
        
        !${pageTitle}^meal.png|thumbnail!
        
        !http://www.lewe.com/wp-content/uploads/2016/03/conf-icon-64.png|conf-icon!
        
        !${pageTitle}^cb-integration-components.png|cb-integration-components.png!
        """.trimIndent(), content )
    }

    /**
     *
     * @param uri
     * @return
     */
    private fun getFileName(uri: String): CompletableFuture<String?> {
        val result = CompletableFuture<String?>()
        try {
            val uriObject = URI.create(uri)
            if ("classpath".equals(uriObject.scheme, ignoreCase = true)) {
                result.completeExceptionally(
                        IllegalArgumentException("'classpath' scheme is not supported!"))
                return result
            }
            val path = Paths.get(uriObject.path)
            result.complete(path.fileName.toString())
        } catch (e: Throwable) {
            result.completeExceptionally(e)
        }
        return result
    }

    @Test
    fun processUrl() {
        val patternUrl = Pattern.compile("(?:(\\$\\{.+\\})\\^)?(.+)")
        run {
            val url = "\${page.title}^image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsEqual.equalTo("\${page.title}"))
            Assert.assertThat(m.group(2), IsEqual.equalTo("image-name.png"))
        }
        run {
            val url = "\${page.title}^./images/image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsEqual.equalTo("\${page.title}"))
            Assert.assertThat(m.group(2), IsEqual.equalTo("./images/image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsEqual.equalTo("image-name.png")) }
                    .join()
        }
        run {
            val url = "image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsEqual.equalTo("image-name.png")) }
                    .join()
        }
        run {
            val url = "./images/image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("./images/image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsEqual.equalTo("image-name.png")) }
                    .join()
        }
        run {
            val url = "file:///image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("file:///image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsEqual.equalTo("image-name.png")) }
                    .join()
        }
        run {
            val url = "http://localhost:8080/image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("http://localhost:8080/image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsEqual.equalTo("image-name.png")) }
                    .join()
        }
        run {
            val url = "classpath:/image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("classpath:/image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable? -> null }
                    .thenAccept { fileName: String? -> Assert.assertThat(fileName, IsNull.nullValue()) }
                    .join()
        }
        run {
            val url = "/image-name.png"
            val m = patternUrl.matcher(url)
            Assert.assertThat(m.matches(), IsEqual.equalTo(true))
            Assert.assertThat(m.groupCount(), IsEqual.equalTo(2))
            Assert.assertThat(m.group(1), IsNull.nullValue())
            Assert.assertThat(m.group(2), IsEqual.equalTo("/image-name.png"))
            getFileName(m.group(2))
                    .exceptionally { ex: Throwable ->
                        Assert.fail(ex.message)
                        null
                    }
                    .thenApply { fileName: String? -> "\${page.title}^$fileName" }
                    .thenAccept { fileName: String -> Assert.assertThat(fileName, IsEqual.equalTo("\${page.title}^image-name.png")) }
                    .join()
        }
    }
}